### PR TITLE
fix: pass hw_kernel_config properly throughout py models

### PR DESCRIPTION
**Pull request by zhangjianning-zjn**

### CODE REVIEW

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1149

**Status:** BLOCKING

**Summary:** P0/0 · P1/0 · P2/1 · P3/1

### Non-blocking Suggestions

#### P2
- **`qwen2_mtp.py` 中 `eh_proj` 未传递 `quant_config`** @ `rtp_llm/models_py/model_desc/qwen2_mtp.py:42`
  - 建议：在 `eh_proj` 调用中补充 `quant_config`：
```python
self.eh_proj = LinearFactory.create_linear_from_weights(
weights.weights[0],
W.multi_tokens_predict_eh_proj,
quant_config=quant_config,
hw_kernel_config=py_hw_kernel_config,
)
```

#### P3
- **`qwen2_mtp.py` 中 `py_hw_kernel_config` 以位置参数传递，风格不一致** @ `rtp_llm/models_py/model_desc/qwen2_mtp.py:62`
  - 建议：改为关键字参数以保持一致：
```python
Qwen3DecoderLayer(
config,
parallelism_config,
idx,
weights.weights[idx],
quant_config=quant_config,
hw_kernel_config=py_hw_kernel_config,
)
```

### Checklist Violations (2 fail / 56 total)

**General Principles Checklist**
- [6.1] Software Engineering — DRY：重复非平凡逻辑被抽取或显式复用  → issue ``qwen2_mtp.py` 中 `eh_proj` 未传递 `quant_config``
  _`Qwen2MtpModel.__init__` 作用域中有 `quant_config` 参数（第 24 行），但 `eh_proj` 创建调用仅传了 `hw_kernel_config` 未传 `quant_config`。对比同 PR 中 `kimi_linear.py` 的 `KimiLinearKDA` 每个 `create_linear_from_weights` 都同时传递 `quant_config=quant_config` 和 `hw_kernel_config=hw_kernel_config`，以及 `MlaAttention`、`CausalAttention` 内部的做法也一致传递两者。若模型以 FP8/INT4 量化部署，`eh_proj` 层将无法选择正确的量化 Linear 实现。该遗漏为此 PR 之前已存在的问题，但既然本 PR 目的是修复参数透传，建议一并修正。_

**RTP-LLM Checklist**
- [I] 代码质量 — 同一功能用统一工具函数  → issue ``qwen2_mtp.py` 中 `py_hw_kernel_config` 以位置参数传递，风格不一致`
  _`Qwen3DecoderLayer` 构造中 `py_hw_kernel_config` 以第 6 个位置参数传递，而本 PR 中其他所有调用点（`kimi_linear.py` 和 `generic_moe.py`）均使用 `hw_kernel_config=...` 关键字参数。位置参数在参数列表扩展时容易静默错位，且同一 PR 内风格不统一降低可读性。_

### Strengths
- 系统性地覆盖了 `generic_moe`、`kimi_linear`、`qwen2_mtp` 三个模型描述器中所有缺失 `hw_kernel_config` 传递的位置，修复完整。
- `GenericMoeLayer.shared_expert_gate` 的旧代码将 `ModelConfig` 对象误传为 `quant_config` 位置参数，本 PR 在添加 `hw_kernel_config` 的同时静默修复了此 bug，改为正确的 `quant_config=quant_config`。
- 大部分调用改用关键字参数传递 `quant_config` 和 `hw_kernel_config`，提升可读性和后续维护安全性。

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1149

**Status:** BLOCKING

**Summary:** P0/0 · P1/1 · P2/0 · P3/0

### Blocking Issues

#### P1
- **shared-expert gate 会使用与权重布局不匹配的 Swizzle 后端** @ `rtp_llm/models_py/model_desc/generic_moe.py:107`
  - 建议：为每个权重按实际布局和 shape 决定后端；shared-expert gate 不满足 Swizzle 约束时应显式回退 no-swizzle，而不是直接传递全局配置。补充 ROCm `use_swizzleA=true` 的模型级回归测试，将该 gate 和完整 MoE 输出与 no-swizzle/Torch reference 对齐。

### Checklist Violations (4 fail / 82 total)

**General Principles Checklist**
- [6.1] Tests — 分布式/跨平台变更有对应覆盖  → issue `shared-expert gate 会使用与权重布局不匹配的 Swizzle 后端`
  _当 `use_swizzleA=true` 且模型含 shared-expert gate 时，新代码把该配置传给 `LinearFactory`，从而选择 `RocmF16LinearWithSwizzle` 并以 `bpreshuffle=True` 执行。但权重重写白名单不含 `W.shared_expert_gate`，且该标量 gate 的输出维度为 1，不满足 Swizzle 的 N%16 约束。因此普通布局权重会被按预重排布局解释，服务可启动但 gate 和最终推理输出错误。_
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `shared-expert gate 会使用与权重布局不匹配的 Swizzle 后端`
  _当 `use_swizzleA=true` 且模型含 shared-expert gate 时，新代码把该配置传给 `LinearFactory`，从而选择 `RocmF16LinearWithSwizzle` 并以 `bpreshuffle=True` 执行。但权重重写白名单不含 `W.shared_expert_gate`，且该标量 gate 的输出维度为 1，不满足 Swizzle 的 N%16 约束。因此普通布局权重会被按预重排布局解释，服务可启动但 gate 和最终推理输出错误。_

**RTP-LLM Checklist**
- [B] 正确性与逻辑 — 传入 kernel/BLAS 的 tensor 物理布局/stride/storage_offset 与契约一致  → issue `shared-expert gate 会使用与权重布局不匹配的 Swizzle 后端`
  _当 `use_swizzleA=true` 且模型含 shared-expert gate 时，新代码把该配置传给 `LinearFactory`，从而选择 `RocmF16LinearWithSwizzle` 并以 `bpreshuffle=True` 执行。但权重重写白名单不含 `W.shared_expert_gate`，且该标量 gate 的输出维度为 1，不满足 Swizzle 的 N%16 约束。因此普通布局权重会被按预重排布局解释，服务可启动但 gate 和最终推理输出错误。_
- [F] 跨平台 — ROCm/XPU/CUDA fallback 的 shape、stride、dtype、sentinel/mask、sampling/beam/logits 数值语义必须与 CUDA/CPU reference 对齐；不支持时必须 graceful reject  → issue `shared-expert gate 会使用与权重布局不匹配的 Swizzle 后端`
  _当 `use_swizzleA=true` 且模型含 shared-expert gate 时，新代码把该配置传给 `LinearFactory`，从而选择 `RocmF16LinearWithSwizzle` 并以 `bpreshuffle=True` 执行。但权重重写白名单不含 `W.shared_expert_gate`，且该标量 gate 的输出维度为 1，不满足 Swizzle 的 N%16 约束。因此普通布局权重会被按预重排布局解释，服务可启动但 gate 和最终推理输出错误。_

### Strengths
- 改动范围集中，并使用关键字参数传递 `quant_config` 与 `hw_kernel_config`，降低位置参数错位风险。
- Qwen2 MTP 的 `eh_proj`、attention 和 MLP 权重均能沿现有配置链获得硬件配置。
- Kimi Linear 的 KDA、MLA、DenseMLP 与 MoE 构造路径实现了统一的配置传播。

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1149

**Status:** BLOCKING

**Summary:** P0/0 · P1/1 · P2/0 · P3/0

### Blocking Issues

#### P1
- **真正受 swizzle 影响的 Dense 和 MTP 路径没有回归测试** @ `rtp_llm/models_py/modules/factory/linear/impl/rocm/test/rocm_linear_test.py:232`
  - 建议：增加 ROCm `use_swizzleA=true` 的聚焦回归用例：通过真实 `LinearFactory` 构造 Generic dense MLP 的 up/down projection 和 Qwen2 MTP 的 `eh_proj`，分别断言选择 swizzle 后端并与 PyTorch 参考结果比较。至少保留一个模型构造级用例，验证配置从模型入口传播到这些线性层。

### Checklist Violations (1 fail / 88 total)

**General Principles Checklist**
- [6.1] Tests — 新逻辑有聚焦单测 + 相关集成/smoke 测试  → issue `真正受 swizzle 影响的 Dense 和 MTP 路径没有回归测试`
  _当 ROCm `use_swizzleA=true` 时，loader 会重排 `W.ffn_w13`、`W.ffn_w2` 和 `W.multi_tokens_predict_eh_proj`；本次也正为 Generic dense MLP 和 Qwen2 MTP 补传配置。但新增用例只实例化 `GenericMoeLayer`，并 mock 掉 `DenseMLP`，也未构造 `Qwen2MtpModel`。因此这两条路径即使仍错误选择无 swizzle 后端，测试也会通过，线上可能发生 GEMM 失败或错误输出。_

### Strengths
- `hw_kernel_config` 通过显式关键字逐层传递，降低了构造参数错位风险。
- shared-expert gate 用例同时验证后端类型和数值结果，能够保护单输出列的无 swizzle 回退。
- shared-expert gate 继续使用原始物理布局，与 loader 未对该权重执行 swizzle 的行为一致。

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1149

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/0 · P3/0

lgtm ready to ci

### Checklist ✅ (88 items passed)

### Strengths
- shared expert gate 针对单输出列显式回退到 no-swizzle backend，避免将不满足布局约束的权重交给 `bpreshuffle` 路径。
- 新增测试同时检查 backend 类型和数值结果，覆盖 Generic MoE dense MLP、shared expert gate 与 Qwen2 MTP `eh_proj`。
- 参数改为关键字传递，降低 `quant_config` 与 `hw_kernel_config` 位置混淆的风险。

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1149

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/0 · P3/0

lgtm ready to ci

### Checklist ✅ (88 items passed)

### Strengths
- `GenericMoeLayer` 对单输出列的 shared-expert gate 显式使用 no-swizzle backend，避免普通布局权重被错误解释。
- Generic dense MLP 与 Qwen2 MTP 测试同时验证 backend 类型和数值结果，覆盖了本次修复的主要 ROCm 路径。
- 新增参数主要采用关键字传递，降低 `quant_config` 与 `hw_kernel_config` 错位风险。

**Comment by wht21** [comment]
internal source has been updated, please review the changes!

**Comment by wht21** [comment]
internal source has been updated, please review the changes!

**Comment by LLLLKKKK** [comment]
## AI Code Review - PR #1149

**Status:** LGTM

**Summary:** P0/0 · P1/0 · P2/0 · P3/0

lgtm ready to ci

### Checklist ✅ (88 items passed)

### Strengths
- `shared_expert_gate` 明确回退到 no-swizzle backend，与该单输出列权重的实际加载布局一致。
- Generic dense MLP 与 Qwen2 MTP 测试同时验证 backend 类型和数值结果，能够捕获配置未透传或布局错配。
- 新增调用主要使用关键字参数，降低 `quant_config` 与 `hw_kernel_config` 发生位置错位的风险。

**Review by LLLLKKKK — APPROVED** [review]
